src/materialized/src/bin/materialized/main.rs

```diff
@@ -335,43 +335,62 @@ fn run() -> Result<(), anyhow::Error> {
-                })
-                .with(FilterLayer::new(
-                    fmt::layer().with_writer(io::stderr),
-                    LevelFilter::WARN,
-                ))
-                .init();
+        match popts.opt_str("log-file").as_deref() {
+            Some("stderr") => {
+                // The user explicitly directed logs to stderr. Log only to stderr
+                // with the user-specified `env_filter`.
+                tracing_subscriber::registry()
+                    .with(env_filter)
+                    .with(fmt::layer().with_writer(io::stderr))
+                    .init()
+            }
+            Some(log_file) => {
+                // The user directed logs to a file. Use the user-specified
+                // `env_filter` to control what gets written to the file. All
+                // output to stderr is silenced.
+                tracing_subscriber::registry()
+                    .with(env_filter)
+                    .with({
+                        let path = PathBuf::from(log_file);
+                        if let Some(parent) = path.parent() {
+                            fs::create_dir_all(parent).with_context(|| {
+                                format!("creating log file directory: {}", parent.display())
+                            })?;
+                        }
+                        let file = fs::OpenOptions::new()
+                            .append(true)
+                            .create(true)
+                            .open(&path)
+                            .with_context(|| format!("creating log file: {}", path.display()))?;
+                        fmt::layer().with_ansi(false).with_writer(move || {
+                            file.try_clone().expect("failed to clone log file")
+                        })
+                    })
+                    .init();
+            }
+            None => {
+                // The user hasn't asked for anything, so emit the full `env_filter` to
+                // configure what goes to the file, but also bubble any warnings or
+                // errors to stderr
+                let path = data_directory.join("materialized.log");
+                let file = fs::OpenOptions::new()
+                    .append(true)
+                    .create(true)
+                    .open(&path)
+                    .with_context(|| format!("creating log file: {}", path.display()))?;
+                tracing_subscriber::registry()
+                    .with(env_filter)
+                    .with(
+                        fmt::layer().with_ansi(false).with_writer(move || {
+                            file.try_clone().expect("failed to clone log file")
+                        }),
+                    )
+                    .with(FilterLayer::new(
+                        fmt::layer().with_writer(io::stderr),
+                        LevelFilter::WARN,
+                    ))
+                    .init();
+            }
```

Comment:
The code here is fiddly enough it'd be great if we could push the conditional down a bit to share more code:

```rust
            Some("stderr") => {
                // The user explicitly directed logs to stderr. Log only to stderr
                // with the user-specified `env_filter`.
                tracing_subscriber::registry()
                    .with(env_filter)
                    .with(fmt::layer().with_writer(io::stderr))
                    .init()
            }
            log_file => {
                // Logging to a file. If the user did not explicitly specify
                // a file, bubble up warnings and errors to stderr.
                let stderr_level = match log_file {
                    Some(_) => LevelFilter::OFF,
                    None => LevelFilter::WARN,
                };
                tracing_subscriber::registry()
                    .with(env_filter)
                    .with({
                        let path = match log_file {
                            Some(log_file) => PathBuf::from(log_file),
                            None => data_directory.join("materialized.log"),
                        };
                        if let Some(parent) = path.parent() {
                            fs::create_dir_all(parent).with_context(|| {
                                format!("creating log file directory: {}", parent.display())
                            })?;
                        }
                        let file = fs::OpenOptions::new()
                            .append(true)
                            .create(true)
                            .open(&path)
                            .with_context(|| format!("creating log file: {}", path.display()))?;
                        fmt::layer().with_ansi(false).with_writer(move || {
                            file.try_clone().expect("failed to clone log file")
                        })
                    })
                    .with(FilterLayer::new(
                        fmt::layer().with_writer(io::stderr),
                        stderr_level,
                    ))
                    .init()
            }
```

Reply:
done.